grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -136,6 +175,122 @@ public static GraqlTraversal createTraversal(PatternAdmin pattern, EmbeddedGrakn
+                    relationshipRolePlayerMap.get(relationship).size() < 2) {
+                iterator.remove();
+            } else {
+                final int[] numRolePlayersHaveType = {0};
+                relationshipRolePlayerMap.get(relationship).forEach(
+                        rolePlayer -> {
+                            if (instanceVarTypeMap.containsKey(rolePlayer)) {
+                                numRolePlayersHaveType[0]++;
+                            }
+                        });
+                if (numRolePlayersHaveType[0] < 2) {
+                    iterator.remove();
+                }
+            }
+        }
+        if (relationshipRolePlayerMap.isEmpty()) return;
+
+        // for each type, get all possible relationship type it could be in
+        Map<Type, Set<RelationshipType>> relationshipMap = new HashMap<>();
+        labelVarTypeMap.values().stream().distinct().forEach(
+                type -> getAllPossibleRelationships(relationshipMap, type));
+
+        relationshipRolePlayerMap.asMap().forEach((relationshipVar, rolePlayerVars) -> {
+            Iterator<Var> rolePlayerVarIterator = rolePlayerVars.iterator();
+
+            Var firstRolePlayer = rolePlayerVarIterator.next();
+            while (!instanceVarTypeMap.containsKey(firstRolePlayer)) {
+                firstRolePlayer = rolePlayerVarIterator.next();
+            }
+
+            Set<Type> possibleRelationShipTypes = getAllPossibleRelationShipTypesOfARolePlayer(
+                    firstRolePlayer, instanceVarTypeMap, relationshipMap);
+            while (rolePlayerVarIterator.hasNext()) {
+                Var nextRolePlayer = rolePlayerVarIterator.next();
+                if (instanceVarTypeMap.containsKey(nextRolePlayer)) {
+
+                    // compute the intersection of possible relationship types
+                    possibleRelationShipTypes.retainAll(getAllPossibleRelationShipTypesOfARolePlayer(
+                            nextRolePlayer, instanceVarTypeMap, relationshipMap));
+                }
+            }
+
+            //TODO: if possibleRelationShipTypes here is empty, the query will not match any data
+            if (possibleRelationShipTypes.size() == 1) {
+
+                Type relationshipType = possibleRelationShipTypes.iterator().next();
+                Label label = relationshipType.getLabel();
+                Var labelVar = var();
+                allFragments.add(Fragments.label(LabelProperty.of(label), labelVar, ImmutableSet.of(label)));
+                allFragments.addAll(EquivalentFragmentSets.isa(IsaProperty.of(label(label).admin()),
+                        relationshipVar, labelVar, relationshipType.isImplicit()).fragments());
+            }
+        });
+    }
+
+    private static Set<Type> getAllPossibleRelationShipTypesOfARolePlayer(
+            Var rolePlayer, Multimap<Var, Type> instanceVarTypeMap, Map<Type, Set<RelationshipType>> relationshipMap) {
```

Comment:
If you move the look-up into `instanceVarTypeMap` outside of this function, the signature can be a bit simpler:
```
private static Set<Type> getAllPossibleRelationshipTypesOfARolePlayer(
    Set<Type> types, Map<Type, Set<RelationshipType>> relationshipMap) {
```